resolver/parallel_best_resolver.go

```diff
@@ -152,89 +134,149 @@ func (r *ParallelBestResolver) Name() string {
 
-	go r1.resolve(request, ch)
+	for _, resolver := range resolvers {
+		logger.WithField("resolver", resolver.resolver).Debug("delegating to resolver")
 
-	logger.WithField("resolver", r2.resolver).Debug("delegating to resolver")
+		go resolver.resolve(request, ch)
+	}
 
-	go r2.resolve(request, ch)
+	response, collectedErrors := evaluateResponses(ctx, logger, ch, resolvers)
+	if response != nil {
+		return response, nil
+	}
+
+	if !r.retryWithDifferentResolver {
+		return nil, fmt.Errorf("resolution was not successful, used resolvers: %s errors: %v",
+			usedResolvers, collectedErrors)
+	}
 
-	for len(collectedErrors) < resolverCount {
-		result := <-ch
+	return r.retryWithDifferent(logger, request, resolvers)
+}
 
-		if result.err != nil {
-			logger.Debug("resolution failed from resolver, cause: ", result.err)
-			collectedErrors = append(collectedErrors, result.err)
-		} else {
-			logger.WithFields(logrus.Fields{
-				"resolver": *result.resolver,
-				"answer":   util.AnswerToString(result.response.Res.Answer),
-			}).Debug("using response from resolver")
+func evaluateResponses(
+	ctx context.Context, logger *logrus.Entry, ch chan requestResponse, resolvers []*upstreamResolverStatus,
+) (*model.Response, []error) {
+	var collectedErrors []error
 
-			return result.response, nil
+	for len(collectedErrors) < len(resolvers) {
+		select {
+		case <-ctx.Done():
+			// this context is currently only set & canceled if resolverCount == 1
+			field := logrus.Fields{"resolver": resolvers[0].resolver}
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				logger.WithFields(field).Debug("upstream exceeded timeout, trying other upstream")
+				resolvers[0].lastErrorTime.Store(time.Now())
+			}
+		case result := <-ch:
+			if result.err != nil {
+				logger.Debug("resolution failed from resolver, cause: ", result.err)
+				collectedErrors = append(collectedErrors, result.err)
+			} else {
+				logger.WithFields(logrus.Fields{
+					"resolver": *result.resolver,
+					"answer":   util.AnswerToString(result.response.Res.Answer),
+				}).Debug("using response from resolver")
+
+				return result.response, nil
+			}
 		}
 	}
 
-	return nil, fmt.Errorf("resolution was not successful, used resolvers: '%s' and '%s' errors: %v",
-		r1.resolver, r2.resolver, collectedErrors)
+	return nil, collectedErrors
+}
+
+func (r *ParallelBestResolver) retryWithDifferent(
+	logger *logrus.Entry, request *model.Request, resolvers []*upstreamResolverStatus,
+) (*model.Response, error) {
+	// second try (if retryWithDifferentResolver == true)
+	resolver := weightedRandom(r.resolvers, resolvers)
+	logger.Debugf("using %s as second resolver", resolver.resolver)
+
+	ch := make(chan requestResponse, 1)
+
+	resolver.resolve(request, ch)
+
+	result := <-ch
+	if result.err != nil {
+		logger.Debug("resolution failed from resolver, cause: ", result.err)
+
+		return nil, errors.New("resolution was not successful, no resolver returned answer in time")
```

Comment:
I think we're better off forwarding the error to not lose that info in release builds:
```suggestion
		return nil, fmt.Errorf("resolution retry failed: %w", err)
```
```